caproto/_log.py

```diff
@@ -143,9 +151,139 @@ def color_logs(color):
 
 
 logger = logging.getLogger('caproto')
+ch_logger = logging.getLogger('caproto.ch')
+search_logger = logging.getLogger('caproto.bcast.search')
 current_handler = None  # overwritten below
 
 
+class logger_nameFilter(logging.Filter):
+    def __init__(self, logger_names):
+        self.logger_names = logger_names
+    def filter(self, record):
+        for i in self.logger_names:
+            if fnmatch.fnmatch(record.name, i):
+                return True
+        return False
+
+
+class PVFilter(logging.Filter):
+    '''
+    Pass through only messages relevant to one or more PV names and
+    pv-unrelated messages.
+
+    Parameters
+    ----------
+    names : string or list of string
+        PV name or PV name list which will be filtered in.
+
+    Returns
+    -------
+    Bool: True or False
+        True if message is not PV related which doesn't have 'pv' as key in extra
+        True if 'pv' as key exists and pv name exists in Filter list.
+        False if message is PV related but pv name isn't in Filter list.
+    '''
+    def __init__(self, names, level='NOTSET', pv_unrelated_flag=True):
+        self.names = names
+        self.level = level
+        self.levelno = logging.getLevelName('level')
+        self.pv_unrelated_flag = pv_unrelated_flag
+
+    def filter(self, record):
+        if hasattr(record, 'pv'):
+            for i in self.names:
+                if fnmatch.fnmatch(record.pv, i) and record.levelno >= self.levelno:
+                    return True
+            return False
+        else:
+            return self.pv_unrelated_flag
+
+
+class PVOnlyFilter(logging.Filter):
+    '''
+    99% as same as PVFilter class but only default return for PV unrelated message
+    Pass through only messages relevant to one or more PV names.
+
+    Parameters
+    ----------
+    names : string or list of string
+        PV name or PV name list which will be filtered in.
+
+    Returns
+    -------
+    Bool: True or False
+        False if message is not PV related which doesn't have 'pv' as key in extra
+        True if 'pv' as key exists and pv name exists in Filter list.
+        False if message is PV related but pv name isn't in Filter list.
+    '''
+    def __init__(self, names, level='NOTSET', pv_unrelated_flag=False):
+        self.names = names
+        self.level = level
+        self.levelno = logging.getLevelName('level')
+        self.pv_unrelated_flag = pv_unrelated_flag
+
+    def filter(self, record):
+        if hasattr(record, 'pv'):
+            for i in self.names:
+                if fnmatch.fnmatch(record.pv, i) and record.levelno >= self.levelno:
+                    return True
+            return False
+        else:
+            return self.pv_unrelated_flag
+
+
+class AddressFilter(logging.Filter):
+    def __init__(self, address_list):
+        self.address_list = address_list
```

Comment:
I think there should be a couple attributes generated based on the passed-in `address_list`:
`self.filtered_address_tuples = {('addr', 5065), ...}`
`self.filtered_addresses = {'addr', ...}`

Then your address check is as simple as `record.address in filtered_address_tuples or record.address[0] in filtered_addresses`, without any string interpolation necessary.

Reply:
I made a similar suggestion in Slack. In `__init__`:
```
self.hosts = []  # list of IP addresses
self.full_addresses = []  # list of (IP, port) tuples
for address in address_list:
    if isinstance(address, str):
        if ':' not in address:
            # Address is like 'XX.XX.XX.XX'
            self.hosts.append(address)
        else:
            # Address is like 'XX.XX.XX.XX:YYYY'
            host, port = address.split(':')
            self.full_addresses.append((host, int(port))
    else:
        # Address is like ('XX.XX.XX.XX', YYYY)
        self.full_addresses.append(address)
```

The advantage is the all of the sorting/normalizing happens once, in `__init__`, and in `filter` all you have to do is loop through `self.full_address` and `self.hosts` to look for a match.

Reply:
Thanks @klauer @danielballan 
I think we may only accept user's pass string for search Or I think in most case passing string is more intuitive. The reason is the thing we print out
`[D 00:37:01.792 _circuit:165] [CLIENT] [192.168.0.10:59545] --->>> [192.168.0.10:5064] (16B) VersionRequest(priority=0, version=13)`

Here we have 
`[192.168.0.10:59545]` instead of `[('192.168.0.10', 59545)]`. If people copy paste, this makes more sense.
If that's true, when the record go inside `fileter`, and there are usually multiple 'target_address'(string) you are interested in and only one `record.address`. So I feel it make more sense to format `record.address` (tuple) to type of `targe_address`, cause you only need do once.

Reply:
> , when the record go inside fileter, and there are usually multiple 'target_address'(string) you are interested in and only one record.address. So I feel it make more sense to format record.address (tuple) to type of targe_address, cause you only need do once.

I think I understand your argument, but I don't think it's correct. Consider that the `__init__` method runs once when the filter is defined by the user. The `filter` method runs many times; it runs each time a record flows through the filter. If we store the target addresses as strings, as you suggest, we have to convert `record.address` (to be precise, actually `record.our_address` and `record.their_address` but that's not important to my argument) from tuple to string every time `filter` runs in order to compare it to the target addresses. If instead, we convert the target addresses to tuple in `__init__`, as @klauer and I suggest, we have to do that exactly once when the filter is defined. Then, every time `filter` runs, no conversion is necessary.